contents/blog/how-we-built-an-app-server.md

```diff
@@ -0,0 +1,300 @@
+
+As you can see by reading the above examples, the index file exports functions that take objects PostHog provides (like configuration or event data). Specifically, our PostHog exports predefined functions that run on conditions like:
+
+- event ingestion (`processEvent`)
+    - useful for modifying incoming data, such as adding geographic data properties or dropping events matching a property filter.
+- every minute, hour, day (`runEveryMinute`, `runEveryHour`, `runEveryDay`)
+    - useful for triggering events based on updates to external services such as Stripe, HubSpot, or GitHub.
+- once processing is complete (`exportEvents`, `onEvent`)
+    - useful for exporting data to external services such as BigQuery, Snowflake, or Databricks.
+
+When users upload apps to PostHog, these we export these functions from the files. From there they run as callable functions on VMs as part of our ingestion pipeline and also have access to extensions like storage, caching, and logging. Overall, it looks like this:
+
+![VM](../images/blog/how-we-built-an-app-server/vm.png)
+
+Structuring apps this way provides flexibility to users while maintaining scalability and security. Added features like scheduling and exporting added new functionality users wanted. Caching and improving the config object (enabling statelessness) helped apps scale effectively. Running apps in VMs helped ensure they were secure and isolated. The next step was to put these apps into production.
+
+## Serving apps and managing data at scale
+
+An app is no good if it doesn't integrate into PostHog. Apps needed to integrate with our ingestion pipeline. They needed to handle an inflow of data, scale, manage tasks, and update themselves. There are many parts needed to make this happen.
+
+Below is a basic diagram of our ingestion pipeline. Most of what we care about is the plugin server and the Kafka data flow leading into it.
+
+```mermaid
+graph TD
+    CLIENT[Client Library]
+    DECIDE["/decide API"]
+    CAPTURE[Capture API]
+    PLUGINS[Plugin server]
+    PERSONS["PostgreSQL (persons table)"]
+    Kafka2[Kafka]
+
+    CLIENT -..-> DECIDE
+
+    CLIENT -..-> CAPTURE
+    CAPTURE --> Kafka
+
+    Kafka <-..- PLUGINS
+
+    PLUGINS <--> PERSONS
+    PLUGINS --> Kafka2
+
+    Kafka2 <-..- ClickHouse
+```
+
+### Using Kafka and Redis
+
+Kafka manages our data flows. It helps us batch, split, and manage parallel work for our apps. For example, Kafka helps us batch and retry events when apps call `processEvent`. We use a Kafka topic (which holds messages and events in a logical order) to continue the flow of data.
+
+Another service we use is Redis to cache data and back the queue for apps. Many parts of the app are reusable and don’t need to be rebuilt each time we run the app. There is data we only need to keep around temporarily. Using Redis to cache data like this dramatically improved app performance and lowered the number of app servers we needed to run.
```

Comment:
> " back the queue for apps."

How? I'm not sure we do.

This line itself reads like "we use caching because caching is good", and is full of vague statements like "many parts of the app" (which ones?), "there is data we need around" (such as?), "data like this" (like what?), "increased app performance" (by how much?).

This reminds me of something from Amazon's rules of writing:

![image](https://user-images.githubusercontent.com/53387/201756978-9dedd906-46f7-4236-86b6-d6816528bbdd.png)

Reply:
Rewrote to get specific example of what we cache in apps (organization details).